api/config/v1/imex.go

```diff
@@ -0,0 +1,102 @@
+/**
+# Copyright 2024 NVIDIA CORPORATION
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+**/
+
+package v1
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+)
+
+const (
+	ImexChannelEnvVar = "NVIDIA_IMEX_CHANNELS"
+)
+
+var errInvalidImexConfig = errors.New("invalid IMEX config")
+
+// Imex stores the configuration options for fabric-attached devices.
+type Imex struct {
+	// ChannelIDs defines a list of channel IDs to inject into containers that request NVIDIA devices.
+	// If a channel ID is specified and the associated channel device node exists, the corresponding
+	// channel will be added to the ContainerAllocateResponse for containers with access to NVIDIA
+	// devices.
+	ChannelIDs []string `json:"channelIDs,omitempty" yaml:"channelIDs,omitempty"`
+	// Required specifies whether the requested IMEX channel IDs are required or not.
+	// If a channel is required, it is expected to exist as the device plugin starts.
+	// If it is not required its injection is skipped if the device nodes do not exist or if its
+	// existence cannot be queried.
+	Required bool `json:"required,omitempty" yaml:"required,omitempty"`
+}
+
+// UnmarshalJSON implements the custom unmarshaler for fields of the IMEX config that require validation.
+func (f *Imex) UnmarshalJSON(b []byte) error {
+	rr := make(map[string]json.RawMessage)
+	if err := json.Unmarshal(b, &rr); err != nil {
+		return err
+	}
+
+	result := Imex{}
+	for k, v := range rr {
+		switch k {
+		case "channelIDs":
+			getValuesChannelIDs := func(r json.RawMessage) ([]string, error) {
+				var strslice []string
+				if err := json.Unmarshal(r, &strslice); err == nil {
+					return strslice, nil
+				}
+				var intslice []int
+				if err := json.Unmarshal(r, &intslice); err == nil {
+					var strslice []string
+					for _, i := range intslice {
+						strslice = append(strslice, fmt.Sprintf("%d", i))
+					}
+					return strslice, nil
+				}
+
+				return nil, fmt.Errorf("unexpected channelIDs: %v", r)
+			}
```

Comment:
Would it be simpler to only accept an intslice?

Reply:
It would probably be simpler, yes. Can we think of a reason to support strings instead?

Reply:
Alternatively should we allow:
```
imex:
  channelIDs: 0
```
in addition to:
```
imex:
  channelIDs: [0]
```

Reply:
Added the ints-only config as a commit on top. I think that is sufficient and definitely simpler.

Reply:
I could even do away with the custom marshaller and check for a valid config in `validateFlags`?

Reply:
```diff --git a/api/config/v1/imex.go b/api/config/v1/imex.go
index a3fd7b5e..ca3050fa 100644
--- a/api/config/v1/imex.go
+++ b/api/config/v1/imex.go
@@ -17,7 +17,6 @@
 package v1
 
 import (
-	"encoding/json"
 	"errors"
 	"fmt"
 )
@@ -42,37 +41,6 @@ type Imex struct {
 	Required bool `json:"required,omitempty" yaml:"required,omitempty"`
 }
 
-// UnmarshalJSON implements the custom unmarshaler for fields of the IMEX config that require validation.
-func (f *Imex) UnmarshalJSON(b []byte) error {
-	rr := make(map[string]json.RawMessage)
-	if err := json.Unmarshal(b, &rr); err != nil {
-		return err
-	}
-
-	result := Imex{}
-	for k, v := range rr {
-		switch k {
-		case "channelIDs":
-			var intslice []int
-			if err := json.Unmarshal(v, &intslice); err != nil {
-				return fmt.Errorf("%w: %v", errInvalidImexConfig, err)
-			}
-			if err := AssertChannelIDsValid(intslice); err != nil {
-				return fmt.Errorf("%w: %v", errInvalidImexConfig, err)
-			}
-			result.ChannelIDs = intslice
-		case "required":
-			var value bool
-			if err := json.Unmarshal(v, &value); err != nil {
-				return err
-			}
-			result.Required = value
-		}
-	}
-	*f = result
-	return nil
-}
-
 // AssertChannelIDsIsValid checks whether the specified list of channel IDs is valid.
 func AssertChannelIDsValid(ids []int) error {
 	switch {
diff --git a/cmd/nvidia-device-plugin/main.go b/cmd/nvidia-device-plugin/main.go
index 6099305b..f6cb1e4a 100644
--- a/cmd/nvidia-device-plugin/main.go
+++ b/cmd/nvidia-device-plugin/main.go
@@ -153,9 +153,6 @@ func main() {
 			Name:    "imex-channel-id",
 			Usage:   "A list of IMEX channels to inject.",
 			EnvVars: []string{"IMEX_CHANNEL_IDS"},
-			Action: func(ctx *cli.Context, s []int) error {
-				return spec.AssertChannelIDsValid(s)
-			},
 		},
 	}
 
@@ -198,6 +195,10 @@ func validateFlags(infolib nvinfo.Interface, config *spec.Config) error {
 		return fmt.Errorf("invalid --device-discovery-strategy option %v", *config.Flags.DeviceDiscoveryStrategy)
 	}
 
+	if err := spec.AssertChannelIDsValid(config.Imex.ChannelIDs); err != nil {
+		return fmt.Errorf("invalid IMEX channel IDs: %w", err)
+	}
+
 	return nil
 }
 
```